### PR TITLE
Gzip request decompress

### PR DESCRIPTION
**Pull request by calebkleveter**
Adds a `NIOHTTPRequestDecompressor` channel handler to decompress incoming HTTP requests.

### Motivation:

There will be times when a client wishes to send larger requests with gzipped bodies to save on network traffic. This PR adds a `NIOHTTPRequestDecompressor` which can be added to the server's channel pipeline so those requests are automatically inflated.

### Modifications:

- Added a `CNIOExtrasZlib_voidPtr_to_BytefPtr` C method.
- Added a `NIOHTTPRequestDecompressor` type.
- Added a `HTTPResponseDecompressorTest` test case.

### Result:

Now you don't have to manually check the `Content-Encoding` header and decompress the body on each incoming request.

### CODE REVIEW

**Comment by weissi** [comment]
@calebkleveter #60  is actually the shared code extraction. Mind checking if you think that'd work for you?

**Comment by weissi** [comment]
@calebkleveter this should now be able to use the functionality from #60

**Comment by calebkleveter** [comment]
@Lukasa I've updated the `NIOHTTPRequestDecompressor` implementation to use `NIOHTTPDecompression`.

File: Sources/NIOHTTPCompression/HTTPRequestDecompressor.swift [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+                    try self.decompressor.decompress(part: &part, buffer: &buffer, originalLength: compression.contentLength)
+
+                    context.fireChannelRead(self.wrapInboundOut(.body(buffer)))
+                } catch let error {
+                    context.fireErrorCaught(error)
+                    return
+                }
+            }
+        case .end(let headers):
+            if self.compression != nil {
```
Comment (Lukasa) [inline Sources/NIOHTTPCompression/HTTPRequestDecompressor.swift]: From a defensive programming perspective, this case should probably `nil` the decoder.

File: Sources/NIOHTTPCompression/HTTPRequestDecompressor.swift [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+                self.decompressor.deinitializeDecoder()
+            }
+
+            context.fireChannelRead(self.wrapInboundOut(.end(headers)))
```
Comment (Lukasa) [inline Sources/NIOHTTPCompression/HTTPRequestDecompressor.swift]: This uses the opposite pattern to line 57, which uses the original `data`. We should be consistent: either change line 57 to be like this or change this line to be like line 57.

File: Tests/NIOExtrasTests/XCTestManifests.swift [hunk 3]
```diff
@@ -0,0 +1,195 @@
+#if !canImport(ObjectiveC)
```
Comment (Lukasa) [inline Tests/NIOExtrasTests/XCTestManifests.swift]: We have a script checked into this repository that generates the tests, at `scripts/generate_linux_tests.rb`. Can you discard these manifests and use that script instead?

**Comment by calebkleveter** [comment]
@Lukasa Fixed. 👍

**Comment by Lukasa** [comment]
@swift-nio-bot test this please

**Pull request merged**